vshard/replicaset.lua

```diff
@@ -75,6 +75,108 @@ local fiber_cond_wait = util.fiber_cond_wait
 local future_wait = util.future_wait
 local gsc = util.generate_self_checker
 
+--
+-- vconnect is asynchronous vshard greeting, saved inside netbox connection.
+-- It stores future object and additional info, needed for its work.
+-- Future is initialized, when the connection is established (inside
+-- netbox_on_connect). The connection cannot be considered "connected"
+-- until vconnect is properly validated.
+--
+local function conn_set_vconnect(conn)
+    assert(conn.replica ~= nil)
+    conn.vconnect = {
+        -- Whether the connection is done to the named replica.
+        is_named = conn.replica.id == conn.replica.name,
+        -- Used to wait for the appearance of vconnect.future object,
+        -- if call is done before the connection is established.
+        future_cond = fiber.cond(),
+    }
+    local vconn = conn.vconnect
+    if not vconn.is_named then
+        -- Skip validation without any waiting.
+        conn.vconnect.is_validated = true
+    end
+end
+
+--
+-- Initialize future object. Should be done, when connection is established
+-- (inside netbox_on_connect).
+--
+local function conn_init_vconnect(conn)
+    local vconn = conn.vconnect
+    if vconn and vconn.is_named then
+        -- Validation should be done after every reconnect.
+        vconn.is_validated = false
+        local opts = {is_async = true}
+        vconn.future = conn:call('vshard.storage._call', {'info'}, opts)
+        vconn.future_cond:broadcast()
+    end
+end
+
+--
+-- Check, that future is ready, and its result is expected, close the connection
+-- if the result is not correct. The function doesn't yield.
+-- @retval true; The correct response is received.
+-- @retval nil, ...; Response is not received or validation error happened.
+--
+local function conn_validate_vconnect(conn)
```

Comment:
`conn_vconnect_check()` (assuming that `conn_vconnect_` is our object-prefix, and that we usually call such validation procedures just `check`).

Reply:
Renamed